File: leet-code/42. Trapping Rain Water/TrappingRainWater42_3.java
package com.example;

public class TrappingRainWater42_3 {

    public int trap(int[] height) {
        int leftMax = height[0], left = 0;
        int rightMax = height[height.length - 1] , right = height.length - 1;
        int trappingWater = 0;
        while (left < right) {
            if (height[left] <= height[right]) {
                left += 1;
                leftMax = Math.max(leftMax, height[left]);
                if (Math.min(leftMax, height[right]) > height[left]) {
                    trappingWater += Math.min(leftMax, height[right]) - height[left];
                }
            } else {
                right -= 1;
                rightMax = Math.max(rightMax, height[right]);
                if (Math.min(height[left], rightMax) > height[right]) {
                    trappingWater += Math.min(height[left], rightMax) - height[right];
                }
            }
        }
        return trappingWater;
    }
}

File: leet-code/42. Trapping Rain Water/TrappingRainWater42_3.kt
package com.example.algorithm

import java.util.LinkedList
import kotlin.math.min

class TrappingRainWater42_3 {
    fun trap(height: IntArray): Int {
        var sums = 0
        val stack = LinkedList<Int>()
        for (i in height.indices) {
            while (stack.isNotEmpty() && height[stack.peek()] < height[i]) {
                val prev = stack.pop()
                if (stack.isEmpty()) {
                    break
                }
                val distance = i - stack.peek() - 1
                val depth = min(height[stack.peek()], height[i]) - height[prev]
                sums += distance * depth
            }
            stack.push(i)
        }
        return sums
    }
}
